### PR TITLE
Bug: `azurerm_frontdoor` fix for caching issue

### LINKED ISSUE #4461: Azure FrontDoor : Caching issue - Unexpected resource modifying
### Terraform (and AzureRM Provider) Version

```sh
terraform -v
Terraform v0.12.9
+ provider.azurerm v1.34.0
```

### Affected Resource(s)

* `azurerm_frontdoor`

### -  --  main.tf
```sh
// Manage Azure FrontDoor
resource "azurerm_frontdoor" "traffic_gateway" {
  name                                         = local.frontdoor_name
  friendly_name                                = local.frontdoor_name
  location                                     = var.location
  resource_group_name                          = var.resource_group
  enforce_backend_pools_certificate_name_check = var.backend_certificate_check # false

  routing_rule {
      name                    = var.routing_rule_name
      accepted_protocols      = ["Https"]
      patterns_to_match       = ["/*"]
      frontend_endpoints      = [var.frontend_endpoint_name]
      forwarding_configuration {
        forwarding_protocol   = var.forwarding_protocol
        backend_pool_name     = var.backend_pool_name
      }
  }

  backend_pool_load_balancing {
    name                            = var.balancing_name
    sample_size                     = local.backand_balancer.sample_size
    successful_samples_required     = local.backand_balancer.successful_samples_required
    additional_latency_milliseconds = local.backend_balancer.additional_latency_milliseconds
  }

  backend_pool_health_probe {
    name                  = var.liveness_probe_name
    path                  = var.healthcheck_api_path
    interval_in_seconds   = 255
    protocol              = "Https"
  }

  backend_pool {
      name            = var.backend_pool_name
      backend {
          host_header = "${var.backend_name}.azure-api.net"
          address     = "${var.backend_name}.azure-api.net"
          http_port   = 80
          https_port  = 443
      }

      load_balancing_name = var.balancing_name
      health_probe_name   = var.liveness_probe_name
  }

  frontend_endpoint {
    name                                      = var.frontend_endpoint_name
    host_name                                 = var.backend_host_name
    web_application_firewall_policy_link_id   = var.waf_enabled != false ? azurerm_frontdoor_firewall_policy.waf_assigned_policy[0].id : null
    custom_https_provisioning_enabled         = false
  }

  tags = {
    environment = var.environment_prefix
  }
}

// Manage FrontDoor firewall policy
resource "azurerm_frontdoor_firewall_policy" "waf_assigned_policy" {
```

### Expected Behavior
```sh 
terraform plan 
OR 
terraform apply
```
**TERRAFORM SHOULD NOTIFY ABOUT ANY CHANGES**

### Actual Behavior

```sh
terraform plan
OR
terraform apply
``` 
 - Operation result: "Apply complete! Resources: 0 added, 0 changed, 0 destroyed." 
 - **UNEXPECTED ACTION: TERRAFORM SILENTLY UPDATED FRONTDOOR WITH ENABLED CACHING**

### Steps to Reproduce

1. `terraform import azurerm_frontdoor.traffic_gateway /subscriptions/...`

2. `terraform apply`

3. Check Caching option on frontDoor servire [Azure Portal]

### References

https://docs.microsoft.com/en-us/azure/frontdoor/front-door-caching

### PR DESCRIPTION
**Pull request by WodansSon**
Original PR #4618
Fixes #4461

### CODE REVIEW

File: website/docs/r/front_door.html.markdown [hunk 1]
```diff
@@ -184,23 +186,23 @@ The `routing_rule` block supports the following:
...
-* `cache_use_dynamic_compression` - (Optional) Whether to use dynamic compression when caching. Valid options are `true` or `false`. Defaults to `true`.
+* `cache_use_dynamic_compression` - (Optional) Whether to use dynamic compression when caching. Valid options are `true` or `false`. Defaults to `false`.
 
-* `cache_query_parameter_strip_directive` - (Optional) Defines cache behavior in releation to query string parameters. Valid options are `StripAll` or `StripNone`. Defaults to `StripNone`
+* `cache_query_parameter_strip_directive` - (Optional) Defines cache behavior in releation to query string parameters. Valid options are `StripAll` or `StripNone`.
 
 * `custom_forwarding_path` - (Optional) Path to use when constructing the request to forward to the backend. This functions as a URL Rewrite. Default behavior preserves the URL path.
 
-* `forwarding_protocol` - (Optional) Protocol to use when redirecting. Valid options are `HttpOnly`, `HttpsOnly`, or `MatchRequest`. Defaults to `MatchRequest`.
+* `forwarding_protocol` - (Optional) Protocol to use when redirecting. Valid options are `HTTPOnly`, `HTTPSOnly`, or `MatchRequest`. Defaults to `HTTPSOnly`.
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply website/docs/r/front_door.html.markdown]: Fixed.

File: azurerm/internal/services/frontdoor/helper.go [hunk 2]
```diff
@@ -201,3 +203,92 @@ func FlattenFrontendEndpointLinkSlice(input *[]frontdoor.FrontendEndpointLink) [
...
+//
+// but in the HealthProbeSettings block the casing of the HealthProbeSettings is (notice the uppercase 'H')::
+// portal-front-door/ -> HealthProbeSettings/healthProbeSettings-1571100669337
+//
+// so if I need to parse the name of the resource from its ID string I would be
+// unable to do so with the current implementation so I normalize the key into
+// a known format so I can reliable parse the ID string.
+//
+// Link to issue: https://github.com/Azure/azure-sdk-for-go/issues/6762
+func ParseAzureResourceIDLowerPath(id string) (*azure.ResourceID, error) {
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/frontdoor/helper.go]: Fixed.

File: azurerm/internal/services/frontdoor/resource_arm_front_door.go [hunk 3]
```diff
@@ -198,7 +200,7 @@ func resourceArmFrontDoor() *schema.Resource {
 											string(frontdoor.HTTPSOnly),
 											string(frontdoor.MatchRequest),
 										}, false),
-										Default: string(frontdoor.MatchRequest),
+										Default: string(frontdoor.HTTPSOnly),
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/frontdoor/resource_arm_front_door.go]: Yes, but the entire cache section it totally broken and this change enables the disabling of the cache. Currently there is no way to disable cache with the originally released resource.

File: azurerm/internal/services/frontdoor/resource_arm_front_door.go [hunk 4]
```diff
@@ -172,24 +173,25 @@ func resourceArmFrontDoor() *schema.Resource {
...
 									},
+									// Remove default value for #4461
 									"cache_query_parameter_strip_directive": {
 										Type:     schema.TypeString,
 										Optional: true,
 										ValidateFunc: validation.StringInSlice([]string{
 											string(frontdoor.StripAll),
 											string(frontdoor.StripNone),
 										}, false),
-										Default: string(frontdoor.StripNone),
```
[the remark this answers is not in the document]
Comment (WodansSon) [reply azurerm/internal/services/frontdoor/resource_arm_front_door.go]: Same.

**Comment by WodansSon** [comment]
@tombuildsstuff @katbyte This PR is 100% will no breaking changes and it would be great if this could make it into v1.42.0 as there are a lot of customers waiting for this fix.

**Comment by WodansSon** [comment]
image

File: website/docs/r/front_door.html.markdown [hunk 5]
```diff
@@ -184,11 +195,13 @@ The `routing_rule` block supports the following:
 
 The `forwarding_configuration` block supports the following:
 
-* `backend_pool_name` - (Required) The name of the Front Door Backend Pool. 
+* `backend_pool_name` - (Required) Specifies the name of the Backend Pool to forward the incoming traffic to. 
+
+* `cache_enabled` - (Optional) Specifies whether to Enable caching or not. Valid options are `true` or `false`. Defaults to `true`.
 
-* `cache_use_dynamic_compression` - (Optional) Whether to use dynamic compression when caching. Valid options are `true` or `false`. Defaults to `true`.
+* `cache_use_dynamic_compression` - (Optional) Whether to use dynamic compression when caching. Valid options are `true` or `false`. Defaults to `false`.
```
Comment (katbyte) [inline website/docs/r/front_door.html.markdown]: should this be true?

File: azurerm/internal/services/frontdoor/resource_arm_front_door.go [hunk 6]
```diff
@@ -1349,21 +1390,35 @@ func flattenArmFrontDoorRoutingRule(input *[]frontdoor.RoutingRule) []interface{
...
 						if dynamicCompression := cacheConfiguration.DynamicCompression; dynamicCompression != "" {
-							if dynamicCompression == frontdoor.DynamicCompressionEnabledEnabled {
-								c["cache_use_dynamic_compression"] = true
+							c["cache_use_dynamic_compression"] = bool(string(dynamicCompression) == string(frontdoor.DynamicCompressionEnabledEnabled))
+						}
+					} else {
+						c["cache_enabled"] = false
+
+						//get `forwarding_configuration`
+						if o, ok := oldByName[*name]; ok {
```
Comment (katbyte) [inline azurerm/internal/services/frontdoor/resource_arm_front_door.go]: Name could be nil so we'll need to nil check it

File: website/docs/r/front_door.html.markdown (hunk 5 above)
Comment (WodansSon) [reply website/docs/r/front_door.html.markdown]: I believe this was a documentation error, as in the schema in master the default value is also false. So I think this is good.

File: azurerm/internal/services/frontdoor/resource_arm_front_door.go (hunk 6 above)
Comment (WodansSon) [reply azurerm/internal/services/frontdoor/resource_arm_front_door.go]: yep yep

**Pull request merged**